File: web/python/driverPlaywright/basePage.py
from playwright.sync_api import Page

class BasePage:
    """
    BasePage class for common page actions.
    """

    def __init__(self, page: Page):
        self.page = page
        self.default_timeout = 30000  # Default timeout in milliseconds

    def do_click(self, selector, timeout=None):
        timeout = timeout or self.default_timeout
        element = self.page.wait_for_selector(selector, state="visible", timeout=timeout)
        element.hover()
        element.click()

    def send_keys(self, selector, text, timeout=None):
        timeout = timeout or self.default_timeout
        element = self.page.wait_for_selector(selector, state="visible", timeout=timeout)
        element.fill(text)

    def navigate(self, url):
        self.page.goto(url, wait_until='domcontentloaded')

    def wait_for_url_to_contain(self, text):
        self.page.wait_for_url(f"**/*{text}*", wait_until='domcontentloaded')

File: web/javascript/driverPlaywright/basePage.js
const { Page } = require('playwright');

class BasePage {
    /**
     * Constructor for BasePage.
     * @param {Page} page - The Playwright page object.
     */
    constructor(page) {
        this.page = page;
        this.defaultTimeout = 30000; // Default timeout in milliseconds
    }

    /**
     * Clicks on an element.
     * @param {ElementHandle} element - The element to click.
     * @param {number} [timeout=this.defaultTimeout] - Optional timeout for waiting.
     */
    async doClick(element, timeout = this.defaultTimeout) {
        await element.waitFor({ state: "visible", timeout });
        await element.hover()
        await element.click();
    }

        /**
     * Clicks on an element.
     * @param {ElementHandle} element - The element to click.
     * @param {number} [timeout=this.defaultTimeout] - Optional timeout for waiting.
     */
        async dblClick(element, timeout = this.defaultTimeout) {
            await element.waitFor({ state: "visible", timeout });
            await element.hover()
            await element.dblclick();
        }    

    /**
     * Sends keystrokes to an element.
     * @param {ElementHandle} element - The element.
     * @param {string} text - The text to type.
     * @param {number} [timeout=this.defaultTimeout] - Optional timeout for waiting.
     */
    async sendKeys(element, text, timeout = this.defaultTimeout) {
        await element.waitFor({ state: "visible", timeout });
        await element.fill(text);
    }

    /**
     * Navigates to a specified URL.
     * @param {string} url - The URL to navigate to.
     */
    async navigate(url) {
        await this.page.goto(url, { waitUntil: 'domcontentloaded' });
    }

     /**
     * Wait for URL to contain
     * @param {string} text - The URL to contain this text.
     */
    async waitForUrlToContain(text) {
        const regex = new RegExp(text);
        await this.page.waitForURL(regex, { waitUntil: 'domcontentloaded' });
    }    
}

module.exports = BasePage;
